src/libbson/src/bson/bson-error.c

```diff
@@ -98,25 +103,69 @@ bson_set_error (bson_error_t *error, /* OUT */
  */
 
 char *
-bson_strerror_r (int err_code,  /* IN */
-                 char *buf,     /* IN */
-                 size_t buflen) /* IN */
+bson_strerror_r (int err_code,                    /* IN */
+                 char *buf BSON_MAYBE_UNUSED,     /* IN */
+                 size_t buflen BSON_MAYBE_UNUSED) /* IN */
 {
    static const char *unknown_msg = "Unknown error";
    char *ret = NULL;
 
 #if defined(_WIN32)
+   // Windows does not provide `strerror_l` or `strerror_r`, but it does
+   // unconditionally provide `strerror_s`.
    if (strerror_s (buf, buflen, err_code) != 0) {
       ret = buf;
    }
-#elif defined(BSON_HAVE_XSI_STRERROR_R)
-   if (strerror_r (err_code, buf, buflen) == 0) {
-      ret = buf;
+#elif defined(__APPLE__)
+   // Apple does not provide `strerror_l`, but it does unconditionally provide
+   // the XSI-compliant `strerror_r`, but only when compiling with Apple Clang.
+   // GNU extensions may still be a problem if we are being compiled with GCC on
+   // Apple. Avoid the compatibility headaches with GNU extensions and the musl
+   // library by assuming QoI will not cause UB when reading the error message
```

Comment:
```suggestion
   // library by assuming implementation will not cause UB when reading the error message
```

I was unaware of QoI abbreviation. Suggestion can be ignored if this is common parlance.